### PR TITLE
Feature for htrun to send ack for correct sync

### PR DESCRIPTION
**Pull request by aashishc1988**
Making a change to htrun module where once it has detected that client has received the right sync, it will send an ack back letting them know that it is okay to run tests. Without this ack, client will continue to loop and eventually time out. Modify README.md to indicate changes made to communication protocol between host and DUT.

### CODE REVIEW

File: mbed_host_tests/host_tests_conn_proxy/conn_proxy.py [hunk 1]
```diff
@@ -269,6 +269,7 @@ def __send_sync(timeout=None):
                             event_queue.put((key, value, time()))
                             idx = sync_uuid_list.index(value)
                             logger.prn_inf("found SYNC in stream: {{%s;%s}} it is #%d sent, queued..."% (key, value, idx))
+                            connector.write_kv("__hstAk","abcd")
```
Comment (cmonr) [inline mbed_host_tests/host_tests_conn_proxy/conn_proxy.py]: Is "abcd" just an arbritrary string?

File: README.md [hunk 2]
```diff
@@ -372,14 +390,15 @@ void GREENTEA_SETUP(const int timeout, const char *host_test_name) {
...
-[1458565465.35][SERI][INF] reset device using 'default' plugin...
-[1458565465.60][SERI][INF] wait for it...
-[1458565466.60][CONN][INF] sending preamble '2f554b1c-bbbf-4b1b-b1f0-f45493282f2c'
-[1458565466.60][SERI][TXD] mbedmbedmbedmbedmbedmbedmbedmbedmbedmbed
-[1458565466.60][SERI][TXD] {{__sync;2f554b1c-bbbf-4b1b-b1f0-f45493282f2c}}
-[1458565466.74][CONN][INF] found SYNC in stream: {{__sync;2f554b1c-bbbf-4b1b-b1f0-f45493282f2c}}, queued...
-[1458565466.74][HTST][INF] sync KV found, uuid=2f554b1c-bbbf-4b1b-b1f0-f45493282f2c, timestamp=1458565466.743000
-[1458565466.74][CONN][RXD] {{__sync;2f554b1c-bbbf-4b1b-b1f0-f45493282f2c}}
+[1530040532.96][GLRM][INF] remote resources reset...
+[1530040535.26][GLRM][TXD] mbedmbedmbedmbedmbedmbedmbedmbedmbedmbed
```
Comment (cmonr) [inline README.md]: Oh, will this still be transmitted? I thought the Mbed OS PR removed it.

File: mbed_host_tests/host_tests_conn_proxy/conn_proxy.py (hunk 1 above)
Comment (aashishc1988) [reply mbed_host_tests/host_tests_conn_proxy/conn_proxy.py]: Yes

File: README.md (hunk 2 above)
Comment (aashishc1988) [reply README.md]: It's still there. I wanted to get rid of it but then I thought may be it was due to some legacy stuff.

File: mbed_host_tests/host_tests_conn_proxy/conn_proxy.py [hunk 3]
```diff
@@ -269,6 +269,7 @@ def __send_sync(timeout=None):
                             event_queue.put((key, value, time()))
                             idx = sync_uuid_list.index(value)
                             logger.prn_inf("found SYNC in stream: {{%s;%s}} it is #%d sent, queued..."% (key, value, idx))
+                            connector.write_kv("__host_ack","abcd")
```
Comment (bridadan) [inline mbed_host_tests/host_tests_conn_proxy/conn_proxy.py]: Should this "abcd" be the sync UUID that was acked? That would be more meaningful to me, even if the device doesn't actually use it.

File: README.md [hunk 4]
```diff
@@ -357,9 +366,18 @@ void GREENTEA_SETUP(const int timeout, const char *host_test_name) {
 	while (1) {
```
Comment (bridadan) [inline README.md]: You added the `runTests = true` bit down below, but I think you'll want to include the `!runTests` condition in this while loop (in addition to the variable declaration itself).

File: README.md [hunk 5]
```diff
@@ -372,14 +390,15 @@ void GREENTEA_SETUP(const int timeout, const char *host_test_name) {
...
-[1458565466.60][SERI][TXD] {{__sync;2f554b1c-bbbf-4b1b-b1f0-f45493282f2c}}
-[1458565466.74][CONN][INF] found SYNC in stream: {{__sync;2f554b1c-bbbf-4b1b-b1f0-f45493282f2c}}, queued...
-[1458565466.74][HTST][INF] sync KV found, uuid=2f554b1c-bbbf-4b1b-b1f0-f45493282f2c, timestamp=1458565466.743000
-[1458565466.74][CONN][RXD] {{__sync;2f554b1c-bbbf-4b1b-b1f0-f45493282f2c}}
+[1530040532.96][GLRM][INF] remote resources reset...
+[1530040535.26][GLRM][TXD] mbedmbedmbedmbedmbedmbedmbedmbedmbedmbed
+[1530040535.26][CONN][INF] sending up to 5 __sync packets (specified with --sync=5)
+[1530040535.26][CONN][INF] sending preamble 'f1ee68c1-86f4-4025-a5af-e755e03802c6'
+[1530040535.26][GLRM][TXD] {{__sync;f1ee68c1-86f4-4025-a5af-e755e03802c6}}
+[1530040536.32][CONN][RXD] Host Key parsed
```
Comment (bridadan) [inline README.md]: If we decide to drop this section (as I commented here: https://github.com/ARMmbed/mbed-os/pull/7340/files#r241550112), then this comment is just a reminder to remove this from the docs as well.

File: README.md (hunk 4 above)
Comment (aashishc1988) [reply README.md]: That is correct. Will fix.

File: mbed_host_tests/host_tests_conn_proxy/conn_proxy.py (hunk 3 above)
Comment (aashishc1988) [reply mbed_host_tests/host_tests_conn_proxy/conn_proxy.py]: Agreed!

File: mbed_host_tests/host_tests_conn_proxy/conn_proxy.py [hunk 6]
```diff
@@ -269,6 +269,7 @@ def __send_sync(timeout=None):
                             event_queue.put((key, value, time()))
                             idx = sync_uuid_list.index(value)
                             logger.prn_inf("found SYNC in stream: {{%s;%s}} it is #%d sent, queued..."% (key, value, idx))
+                            connector.write_kv("__host_ack",sync_uuid)
```
Comment (bridadan) [inline mbed_host_tests/host_tests_conn_proxy/conn_proxy.py]: Pretty sure you mean `value` instead of `sync_uuid` here.
```suggestion
                            connector.write_kv("__host_ack", value)
```

File: README.md [hunk 7]
```diff
@@ -321,7 +322,8 @@ Hanshake between DUT and host is a sequence of ```__sync``` events send between
 After reset:
 * DUT calls function ```GREENTEA_SETUP(timeout, "host test name");``` which
 * calls immediately ```greentea_parse_kv``` (blocking parse of input serial port for event ```{{__sync;UUID}}```).
-* When ```__sync``` packet is parsed in the stream DUT sends back (echoes) ```__sync``` event with the same [UUID](https://en.wikipedia.org/wiki/Universally_unique_identifier#Version_3_.28MD5_hash_.26_namespace.29) as payload. UUID is a random value e.g.  ```5f8dbbd2-199a-449c-b286-343a57da7a37```.
+* When ```__sync``` packet is parsed in the stream DUT sends back (echoes) ```__sync``` event with the same [UUID](https://en.wikipedia.org/wiki/Universally_unique_identifier#Version_3_.28MD5_hash_.26_namespace.29) as payload. UUID is a random value e.g.  ```5f8dbbd2-199a-449c-b286-343a57da7a37```. Once it sends back ```__sync``` event, it waits for an acknowledgment from Host ```{{__host_ack;abcd}}```) which indicates that the sync key received was correct. "abcd" here
```
Comment (bridadan) [inline README.md]: Just now noticed this, sorry for the extra comments! I'd recommend changing this section:

> Once it sends back ```__sync``` event, it waits for an acknowledgment from Host ```{{__host_ack;abcd}}```) which indicates that the sync key received was correct. "abcd" here represents a random string that DUT/host does not care about.

to this:

> Once it sends back ```__sync``` event, it waits for an acknowledgment from Host ```{{__host_ack;5f8dbbd2-199a-449c-b286-343a57da7a37}}```) which indicates that the sync key received was correct.

File: README.md [hunk 8]
```diff
@@ -342,7 +343,14 @@ greentea_parse_kv(key,value)   |                   |
...
 echoes __sync event with       |------------------>|
 the same UUID to master        |                   |
+                               |                   |self.send_kv("__host_ack", UUID)
+                               |{{__host_ack;UUID}}|<-----------------------------
+                               |<------------------|
+                               |                   |
                                |                   |
+greentea_parse_kv              |{{__host_ack;UUID}}|
+echoes __host_ack event        |------------------>|
+to master                      |                   |
```
Comment (theotherjimmy) [inline README.md]: While this may happen, it's not implemented in this PR or required. Perhaps we should simply drop it from this diagram?

Comment (aashishc1988) [reply README.md]: Actually come to think of it, we should drop all the changes in readme for this PR and add them with the greentea PR. If you think of it, nothing in this PR affects how greentea and htrun works. Thoughts?

Comment (bridadan) [reply README.md]: I think its fine for most of it to stay here, just because that's where its historically been. The greentea-client inside the OS doesn't mention all the different keys and such.

**Review by theotherjimmy — APPROVED** [review]

**Review by bridadan — APPROVED** [review]

**Review by bridadan — CHANGES_REQUESTED** [review]
Ohhh hmm. I remember a corner case now. Certain tests may not check the first key being received to see if it is what the expect. For instance: https://github.com/ARMmbed/mbed-os/blob/1e676f6eda27c84f61fa97626b1980b9e410b769/TESTS/mbed_hal/rtc_time_conv/main.cpp#L144

This test will assert if any key that is not equal to "passed" is sent after the test starts. We'll double checkt the mbed-os suite of tests and see how to rest of them behave with this change.

**Review by bridadan — APPROVED** [review]
@aashishc1988 tested this offline and it appears the Mbed OS test suite is unaffected.

**Review by bridadan — CHANGES_REQUESTED** [review]
Hey @aashishc1988, I was able to cause a failure to show up using the `NRF51_DK` (that has flow control enabled). I pulled in your patch of htrun on my system and I'm using the `master` branch of Mbed OS. This is the test command: `mbed test -m NRF51_DK -t GCC_ARM -v -n tests-mbed_hal-rtc_time_conv`

And this is the output:
```
[mbed] Exec "mbedgt --test-spec .\BUILD\tests\NRF51_DK\GCC_ARM\test_spec.json -n tests-mbed_hal-rtc_time_conv -V" in "C:\Users\bridan01\onedrive_arm\Documents\dev\mbed-os"
mbedgt: greentea test automation tool ver. 1.5.0
mbedgt: test specification file '.\BUILD\tests\NRF51_DK\GCC_ARM\test_spec.json' (specified with --test-spec option)
mbedgt: using '.\BUILD\tests\NRF51_DK\GCC_ARM\test_spec.json' from current directory!
mbedgt: detecting connected mbed-enabled devices...
mbedgt: detected 1 device
        | platform_name | platform_name_unique | serial_port | mount_point | target_id                                        |
        |---------------|----------------------|-------------|-------------|--------------------------------------------------|
        | NRF51_DK      | NRF51_DK[0]          | COM40       | D:          | 110000005032312036393738313037313032303197969903 |
mbedgt: processing target 'NRF51_DK' toolchain 'GCC_ARM' compatible platforms... (note: switch set to --parallel 1)
        | platform_name | platform_name_unique | serial_port | mount_point | target_id                                        |
        |---------------|----------------------|-------------|-------------|--------------------------------------------------|
        | NRF51_DK      | NRF51_DK[0]          | COM40:9600  | D:          | 110000005032312036393738313037313032303197969903 |
mbedgt: test case filter (specified with -n option)
        tests-mbed_hal-rtc_time_conv
        test filtered in 'tests-mbed_hal-rtc_time_conv'
mbedgt: running 1 test for platform 'NRF51_DK' and toolchain 'GCC_ARM'
        use 1 instance of execution threads for testing
mbedgt: checking for 'host_tests' directory above image directory structure
        found 'host_tests' directory in: 'TESTS\host_tests'
mbedgt: selecting test case observer...
        calling mbedhtrun: mbedhtrun -m NRF51_DK -p COM40:9600 -f "BUILD/tests/NRF51_DK/GCC_ARM/TESTS/mbed_hal/rtc_time_conv/rtc_time_conv.hex" -e "TESTS\host_tests" -d D: -c default -t 110000005032312036393738313037313032303197969903 -r default -C 6 --sync 5 -P 60
mbedgt: mbed-host-test-runner: started
[1546469940.57][HTST][INF] host test executor ver. 1.4.4
[1546469940.57][HTST][INF] copy image onto target...
[1546469940.58][COPY][INF] Waiting up to 60 sec for '110000005032312036393738313037313032303197969903' mount point (current is 'D:')...
        1 file(s) copied.
[1546469966.14][HTST][INF] starting host test process...
[1546469967.18][CONN][INF] starting connection process...
[1546469967.18][CONN][INF] notify event queue about extra 60 sec timeout for serial port pooling
[1546469967.18][CONN][INF] initializing serial port listener...
[1546469967.18][PLGN][INF] Waiting up to 60 sec for '110000005032312036393738313037313032303197969903' serial port (current is 'COM40')...
[1546469967.19][HTST][INF] setting timeout to: 60 sec
[1546469967.27][SERI][INF] serial(port=COM40, baudrate=9600, read_timeout=0.01, write_timeout=5)
[1546469967.27][SERI][INF] reset device using 'default' plugin...
[1546469967.52][SERI][INF] waiting 1.00 sec after reset
[1546469968.52][SERI][INF] wait for it...
[1546469968.52][SERI][TXD] mbedmbedmbedmbedmbedmbedmbedmbedmbedmbed
[1546469968.52][CONN][INF] sending up to 5 __sync packets (specified with --sync=5)
[1546469968.52][CONN][INF] sending preamble 'de09b4cb-fbdb-4112-bc79-9c3e24fb5ad0'
[1546469968.52][SERI][TXD] {{__sync;de09b4cb-fbdb-4112-bc79-9c3e24fb5ad0}}
[1546469968.66][CONN][RXD] mbedmbedmbedmbedmbedmbedmbedmbed
[1546469968.71][CONN][INF] found SYNC in stream: {{__sync;de09b4cb-fbdb-4112-bc79-9c3e24fb5ad0}} it is #0 sent, queued...
[1546469968.72][SERI][TXD] {{__host_ack;de09b4cb-fbdb-4112-bc79-9c3e24fb5ad0}}
[1546469968.72][HTST][INF] sync KV found, uuid=de09b4cb-fbdb-4112-bc79-9c3e24fb5ad0, timestamp=1546469968.715000
[1546469968.74][CONN][INF] found KV pair in stream: {{__version;1.3.0}}, queued...
[1546469968.74][HTST][INF] DUT greentea-client version: 1.3.0
[1546469968.75][CONN][INF] found KV pair in stream: {{__timeout;300}}, queued...
[1546469968.75][HTST][INF] setting timeout to: 300 sec
[1546469968.79][CONN][INF] found KV pair in stream: {{__host_test_name;rtc_calc_auto}}, queued...
[1546469968.79][HTST][INF] host test class: '<class 'rtc_calc_auto.RTC_time_calc_test'>'
[1546469968.79][HTST][INF] host test setup() call...
[1546469968.79][HTST][INF] CALLBACKs updated
[1546469968.79][HTST][INF] host test detected: rtc_calc_auto
[1546469968.81][CONN][INF] found KV pair in stream: {{__testcase_count;2}}, queued...
[1546469968.85][CONN][RXD] >>> Running 2 test cases...
[1546469968.93][CONN][INF] found KV pair in stream: {{__testcase_name;test make time and local time - RTC leap years full support}}, queued...
[1546469969.02][CONN][RXD]
[1546469969.02][CONN][INF] found KV pair in stream: {{__testcase_name;test make time and local time - RTC leap years partial support}}, queued...
[1546469969.11][CONN][RXD] >>> Running case #1: 'test make time and local time - RTC leap years full support'...
[1546469969.19][CONN][INF] found KV pair in stream: {{__testcase_start;test make time and local time - RTC leap years full support}}, queued...
[1546469969.22][CONN][INF] found KV pair in stream: {{leap_year_setup;0}}, queued...
[1546469969.24][CONN][INF] found KV pair in stream: {{timestamp;82800}}, queued...
[1546469969.25][SERI][TXD] {{passed;262144}}
mbedgt: :144::FAIL: Expected 'passed' Was '__host_ack'
[1546469969.34][CONN][RXD] :144::FAIL: Expected 'passed' Was '__host_ack'
[1546469969.43][CONN][INF] found KV pair in stream: {{__testcase_finish;test make time and local time - RTC leap years full support;0;1}}, queued...
[1546469969.56][CONN][RXD] >>> 'test make time and local time - RTC leap years full support': 0 passed, 1 failed with reason 'Assertion Failed'
[1546469969.56][CONN][RXD]
[1546469969.62][CONN][RXD] >>> Test cases: 0 passed, 1 failed with reason 'Assertion Failed'
[1546469969.64][CONN][RXD] >>> TESTS FAILED!
[1546469969.67][CONN][INF] found KV pair in stream: {{__testcase_summary;0;1}}, queued...
[1546469969.70][CONN][INF] found KV pair in stream: {{end;failure}}, queued...
[1546469969.70][HTST][INF] __notify_complete(False)
[1546469969.70][HTST][INF] __exit_event_queue received
[1546469969.70][HTST][INF] test suite run finished after 0.95 sec...
[1546469969.71][CONN][INF] found KV pair in stream: {{__exit;0}}, queued...
[1546469969.71][CONN][INF] received special event '__host_test_finished' value='True', finishing
[1546469969.72][HTST][INF] CONN exited with code: 0
[1546469969.72][HTST][INF] Some events in queue
[1546469969.72][HTST][INF] stopped consuming events
[1546469969.72][HTST][INF] host test result() call skipped, received: False
[1546469969.72][HTST][WRN] missing __exit event from DUT
[1546469969.72][HTST][INF] calling blocking teardown()
[1546469969.72][HTST][INF] teardown() finished
[1546469969.72][HTST][INF] {{result;failure}}
mbedgt: retry mbedhtrun 1/1
mbedgt: ['mbedhtrun', '-m', u'NRF51_DK', '-p', u'COM40:9600', '-f', u'"BUILD/tests/NRF51_DK/GCC_ARM/TESTS/mbed_hal/rtc_time_conv/rtc_time_conv.hex"', '-e', u'"TESTS\\host_tests"', '-d', 'D:', '-c', 'default', '-t', u'110000005032312036393738313037313032303197969903', '-r', 'default', '-C', '6', '--sync', '5', '-P', '60'] failed after 1 count
mbedgt: checking for GCOV data...
mbedgt: mbed-host-test-runner: stopped and returned 'FAIL'
mbedgt: test on hardware with target id: 110000005032312036393738313037313032303197969903
mbedgt: test suite 'tests-mbed_hal-rtc_time_conv' .................................................... FAIL in 30.42 sec
        test case: 'test make time and local time - RTC leap years full support' ..................... FAIL in 0.24 sec
        test case: 'test make time and local time - RTC leap years partial support' .................. SKIPPED in 0.00 sec
mbedgt: test case summary: 0 passes, 1 failure
mbedgt: utest test case summary mismatch: utest reported passes and failures miscount!
        reported by utest: passes = 0, failures 1)
        test case result count: passes = 0, failures 2)
mbedgt: all tests finished!
mbedgt: shuffle seed: 0.7684929992
mbedgt: test suite report:
| target           | platform_name | test suite                   | result | elapsed_time (sec) | copy_method |
|------------------|---------------|------------------------------|--------|--------------------|-------------|
| NRF51_DK-GCC_ARM | NRF51_DK      | tests-mbed_hal-rtc_time_conv | FAIL   | 30.42              | default     |
mbedgt: test suite results: 1 FAIL
mbedgt: test case report:
| target           | platform_name | test suite                   | test case                                                      | passed | failed | result  | elapsed_time (sec) |
|------------------|---------------|------------------------------|----------------------------------------------------------------|--------|--------|---------|--------------------|
| NRF51_DK-GCC_ARM | NRF51_DK      | tests-mbed_hal-rtc_time_conv | test make time and local time - RTC leap years full support    | 0      | 1      | FAIL    | 0.24               |
| NRF51_DK-GCC_ARM | NRF51_DK      | tests-mbed_hal-rtc_time_conv | test make time and local time - RTC leap years partial support | 0      | 0      | SKIPPED | 0.0                |
mbedgt: test case results: 1 FAIL / 1 SKIPPED
mbedgt: completed in 44.15 sec
mbedgt: exited with code 1
[mbed] ERROR: "mbedgt" returned error.
       Code: 1
       Path: "C:\Users\bridan01\onedrive_arm\Documents\dev\mbed-os"
       Command: "mbedgt --test-spec .\BUILD\tests\NRF51_DK\GCC_ARM\test_spec.json -n tests-mbed_hal-rtc_time_conv -V"
       Tip: You could retry the last command with "-v" flag for verbose output
---
```

This is the bad line: `mbedgt: :144::FAIL: Expected 'passed' Was '__host_ack'`

**Comment by bridadan** [comment]
Note the test `tests-mbed_platform-system_reset` also failed with the following assert:

```
[1546466828.93][SERI][TXD] {{reset;0}}
mbedgt: :41::FAIL: Unexpected message key.
[1546466829.01][CONN][RXD] :41::FAIL: Unexpected message key.
```

**Comment by bridadan** [comment]
Due to some of the issues we've discovered while investigating this, I'll close this particular PR (since this repo is moving! See #238). We can revisit this issue in https://github.com/armmbed/mbed-os-tools.
